meta-client/src/main/java/org/apache/helix/metaclient/impl/zk/adapter/ChildListenerAdapter.java

```diff
@@ -39,7 +39,7 @@ public ChildListenerAdapter(ChildChangeListener listener) {
   private static ChildChangeListener.ChangeType convertType(Watcher.Event.EventType eventType) {
     switch (eventType) {
       case NodeCreated: return ChildChangeListener.ChangeType.ENTRY_CREATED;
-      case NodeChildrenChanged: return ChildChangeListener.ChangeType.ENTRY_DATA_CHANGE;
```

Comment:
Is my understanding correct that persistent and recursive watcher here and event "NodeChildrenChanged" in helix manager (callback handler events) are totally different? What I'm trying to understand is after persistent recursive watcher change we would get "NodeDataChanged" event instead of "NodeChildrenChanged" event but i can see in helix manager we still fire events of type "NodeChildrenChanged".

Reply:
Yes. The event type we fire here and what helix manager hire is totally different. 
We add APIs for persistent recursive listener in ZkClient here. We need to redesign the callback handling logic in helix manager to adopt persistent recursive listener.